include/deal.II/base/symbolic_function.h

```diff
@@ -0,0 +1,315 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2019 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_symbolic_function_h
+#define dealii_symbolic_function_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/exceptions.h>
+#include <deal.II/base/function.h>
+#include <deal.II/base/point.h>
+#include <deal.II/base/subscriptor.h>
+#include <deal.II/base/symmetric_tensor.h>
+#include <deal.II/base/tensor.h>
+
+#include <deal.II/differentiation/sd.h>
+
+#include <functional>
+#include <iostream>
+#include <vector>
+
+DEAL_II_NAMESPACE_OPEN
+
+// Forward declarations
+#ifndef DOXYGEN
+template <typename number>
+class Vector;
+namespace Functions
+{
+  template <int dim, typename RangeNumberType>
+  class SymbolicFunction;
+}
+#endif
+
+namespace Functions
+{
+#ifdef DEAL_II_WITH_SYMENGINE
```

Comment:
Can you update the file `doc/doxygen/options.dox.in` to also always `#define` this variable so that the documentation shows up?

To make sure people get reasonable error messages when using this class but not having SymEngine installed, it's probably useful to have
```
  #else

  template <int dim>
  class SymbolicFunction {
    static_assert (false, "You can't use this class without having SymEngine installed and detected during "
                                    "configuration of deal.II");
  };
```

Reply:
> Can you update the file doc/doxygen/options.dox.in to also always #define this variable so that the documentation shows up?

It is already there.... Is the documentation not showing up?

Reply:
As far as the static assert is concerned, we don't do this for any of our other classes, which are instead simply forward declared (as in this case, at the beginning of the header).

Moreover, this would break the include tests, which try to include each .h file and compile with it...

Reply:
Ah, I hadn't checked `options.dox.in` :-(

We have quite a lot of classes that have a stub for the case where a dependency is not met. `p::d::Triangulation` is an example. But I don't see why the `static_assert` breaks the test: you can compile a file, you just can't instantiate the class template. Or do I miss something?

Reply:
MWE:

```
class MyClass {
static_assert(false, "Error message");
}

int main () {
return 0;
}
```
does not compile for me:
```
../snippet.cc:2:1: error: static_assert failed "Error message"
static_assert(false, "Error message");
^             ~~~~~
1 error generated.
ninja: build stopped: subcommand failed.
```

Same with putting the assert in the constructor.

Reply:
The file, with your suggestion, cannot be included.

Reply:
Hm :-] Then maybe make it a regular assertion, or make the assertion template-argument dependent. Or just forget about it -- it's not that important.

Reply:
Done.